kernel/sched.c

```diff
@@ -1752,48 +1759,49 @@ void z_thread_abort(struct k_thread *thread)
 	}
 
 #ifdef CONFIG_SMP
-	if (is_aborting(thread) && thread == _current && arch_is_in_isr()) {
-		/* Another CPU is spinning for us, don't deadlock */
-		end_thread(thread);
-	}
+	if (is_aborting(_current)) {
+		/*
+		 * Another CPU may be spinning in an ISR waiting for the
+		 * the current thread to abort. Abort it now to help avoid
+		 * a potential deadlock.
+		 */
+		end_thread(_current);
 
-	bool active = thread_active_elsewhere(thread);
+		if (!arch_is_in_isr()) {
+			z_swap(&sched_spinlock, key);
+			__ASSERT(false, "aborted _current back from dead");
```

Comment:
So... this isn't a simplification anymore.  This is adding code to handle an extra case, which I guess is an optimization which catches the "aborted from another CPU" earlier, in the case where a thread is also self-aborting at the same time.    I guess I don't see the value there?   Why do we want to add code byte to make that particular operation faster?

And it seems needless to me anyway?   How would you reasonably hit this in practice?  The "aborting" state is flagged synchronously with the other CPU sending an IPI, this is only hit in the case where the current thread is managing to get into an (unlocked) abort call before the IPI arrives.  That seems almost impossible to hit in practice.

Really, can we just remove this first patch?  I honestly think it's hurting and not helping.  For sure, 100%, this is subtle code and takes time to understand.  But right now the changes mostly amount to your exploratory process of understanding and not any improvement on what's there.

Reply:
I am okay with dropping that commit.